server/app/filters/ApiKeyUsageFilter.java

```diff
@@ -0,0 +1,74 @@
+package filters;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import auth.ProfileUtils;
+import java.time.Duration;
+import java.util.Optional;
+import java.util.concurrent.Executor;
+import javax.inject.Inject;
+import javax.inject.Provider;
+import play.api.libs.concurrent.AkkaSchedulerProvider;
+import play.mvc.EssentialAction;
+import play.mvc.EssentialFilter;
+import play.mvc.Http;
+import scala.concurrent.ExecutionContext;
+import services.apikey.ApiKeyService;
+
+/**
+ * This filter looks for requests with paths that begin with /api and record the usage data for the
+ * relevant API key if it has one. The API key update is scheduled to run in a background through to
+ * reduce latency and to ensure issues with recording the usage do not cause API requests to fail.
+ */
+public class ApiKeyUsageFilter extends EssentialFilter {
+
+  private final AkkaSchedulerProvider akkaSchedulerProvider;
+  private final Provider<ApiKeyService> apiKeyServiceProvider;
+  private final Executor exec;
+  private final Provider<ProfileUtils> profileUtilsProvider;
+
+  @Inject
+  public ApiKeyUsageFilter(
+      AkkaSchedulerProvider akkaSchedulerProvider,
+      Provider<ApiKeyService> apiKeyServiceProvider,
+      Executor exec,
+      Provider<ProfileUtils> profileUtilsProvider) {
+    this.akkaSchedulerProvider = checkNotNull(akkaSchedulerProvider);
+    this.apiKeyServiceProvider = checkNotNull(apiKeyServiceProvider);
+    this.exec = checkNotNull(exec);
+    this.profileUtilsProvider = checkNotNull(profileUtilsProvider);
+  }
+
+  @Override
+  public EssentialAction apply(EssentialAction next) {
+    return EssentialAction.of(
+        (Http.RequestHeader request) ->
+            next.apply(request)
+                .map(
+                    result -> {
+                      if (request.path().startsWith("/api")) {
+                        Optional<String> maybeApiKeyId =
+                            profileUtilsProvider.get().currentApiKeyId(request);
+
+                        // If the key ID is not present then the request was not
+                        // authenticated and does not need to be recorded.
+                        if (maybeApiKeyId.isPresent()) {
+                          String remoteAddress = request.remoteAddress();
+
+                          akkaSchedulerProvider
+                              .get()
+                              .scheduleOnce(
```

Comment:
Not blocking: Probably complete overkill and not necessary here, but it might make sense to debounce / batch requests here to reduce database contention when there's a large influx of API requests.

That'd look something like having some sort of background-thread-local-variable that's incremented here, then scheduling a task (that's debounced every 1s or so) to update the current state. It could then:
1) Lock
2) Consult the background-thread-local-variable and increment by that amount
3) Unlock

Reply:
That'd definitely be an improvement, but I don't think any environment is going to see enough traffic to make the complexity tradeoff worth it. It's also ok if the count isn't perfectly accurate

Reply:
Sounds good to me. Can always add this if the need arises.